java/arcs/core/data/proto/RecipeProtoDecoder.kt

```diff
@@ -17,12 +17,21 @@ import arcs.core.data.Recipe
 /** Converts a [RecipeProto] into [Recipe]. */
 fun RecipeProto.decode(particleSpecs: Map<String, ParticleSpec>): Recipe {
     val recipeHandles = mutableMapOf<String, Recipe.Handle>()
-    handlesList.forEach {
-        val oldValue = recipeHandles.put(it.name, it.decode())
+    val decodeAndUpdate = { it: HandleProto ->
+        val oldValue = recipeHandles.put(it.name, it.decode(recipeHandles))
         require(oldValue == null) {
             "Duplicate handle '${it.name}' when decoding recipe '$name'."
         }
     }
+
+    handlesList
+        .filter { it.fate.decode() != Recipe.Handle.Fate.JOIN }
+        .forEach(decodeAndUpdate)
+
+    handlesList
+        .filter { it.fate.decode() == Recipe.Handle.Fate.JOIN }
+        .forEach(decodeAndUpdate)
```

Comment:
Curious why these need to be separate? Comments would be helpful to explain why.

Reply:
+1

Reply:
Associated handles appear only in with `join` fates. In the modified handle decode, we associate the handles via map lookup and assert when they're not there. 

Breaking out the recipe decoding into two passes means that the referred-to handles will be in the map by the time they are referenced. 

A corner case that just came to mind: can you join a join handle?